mllib/src/main/scala/org/apache/spark/ml/feature/Imputer.scala

```diff
@@ -0,0 +1,290 @@
+    require(!schema.fieldNames.contains($(outputCol)),
+      s"Output column ${$(outputCol)} already exists.")
+    val outputFields = schema.fields :+ StructField($(outputCol), new VectorUDT, false)
+    StructType(outputFields)
+  }
+
+}
+
+/**
+ * :: Experimental ::
+ * Imputation estimator for completing missing values, either using the mean, the median or
+ * the most frequent value of the column in which the missing values are located. This class
+ * also allows for different missing values.
+ */
+@Experimental
+class Imputer @Since("2.0.0")(override val uid: String)
+  extends Estimator[ImputerModel] with ImputerParams with DefaultParamsWritable {
+
+  @Since("2.0.0")
+  def this() = this(Identifiable.randomUID("imputer"))
+
+  /** @group setParam */
+  def setInputCol(value: String): this.type = set(inputCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  /**
+   * Imputation strategy. Available options are "mean", "median" and "most".
+   * @group setParam
+   */
+  def setStrategy(value: String): this.type = set(strategy, value)
+
+  /** @group setParam */
+  def setMissingValue(value: Double): this.type = set(missingValue, value)
+
+  setDefault(strategy -> "mean", missingValue -> Double.NaN)
+
+  override def fit(dataset: DataFrame): ImputerModel = {
+    val alternate = dataset.select($(inputCol)).schema.fields(0).dataType match {
+      case DoubleType =>
+        Vectors.dense(getColStatistics(dataset, $(inputCol)))
+      case _: VectorUDT =>
+        val vl = dataset.first().getAs[Vector]($(inputCol)).size
+        val statisticsArray = new Array[Double](vl)
+        (0 until vl).foreach(i => {
+          val getI = udf((v: Vector) => v(i))
+          val tempColName = $(inputCol) + i
+          val tempData = dataset.where(s"${$(inputCol)} IS NOT NULL")
+            .select($(inputCol)).withColumn(tempColName, getI(col($(inputCol))))
+          statisticsArray(i) = getColStatistics(tempData, tempColName)
+        })
+        Vectors.dense(statisticsArray)
+    }
+    copyValues(new ImputerModel(uid, alternate).setParent(this))
+  }
+
+  /** Extract the statistics info from a Double column according to the strategy */
+  private def getColStatistics(dataset: DataFrame, colName: String): Double = {
+    val missValue = $(missingValue) match {
+      case Double.NaN => "NaN"
+      case _ => $(missingValue).toString
+    }
+    val filteredDF = dataset.select(colName).where(s"$colName != '$missValue'")
+    val colStatistics = $(strategy) match {
+      case "mean" =>
+        filteredDF.selectExpr(s"avg($colName)").first().getDouble(0)
+      case "median" =>
+        // TODO: optimize the sort with quick-select or Percentile(Hive) if required
+        val rddDouble = filteredDF.rdd.map(_.getDouble(0))
+        rddDouble.sortBy(d => d).zipWithIndex().map {
+          case (v, idx) => (idx, v)
+        }.lookup(rddDouble.count() / 2).head
+      case "most" =>
```

Comment:
I'm a little worried about performance here - on huge columns of `Double` this could be a problem. I struggle to actually see a use case for `most`... perhaps on categorical or ordinal columns it can make sense. 

Will think about it a bit more.